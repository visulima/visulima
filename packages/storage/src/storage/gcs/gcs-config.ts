const GCSConfig = {
    authScopes: [
        "https://www.googleapis.com/auth/iam",
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/devstorage.full_control",
    ],
    storageAPI: "https://storage.googleapis.com/storage/v1/b",
    uploadAPI: "https://storage.googleapis.com/upload/storage/v1/b",
};

export default GCSConfig;
