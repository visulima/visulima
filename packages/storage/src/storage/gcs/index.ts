export { default as GCSConfig } from "./gcs-config";
export { default as GCSFile } from "./gcs-file";
export { default as GCSMetaStorage } from "./gcs-meta-storage";
export { default as GCStorage } from "./gcs-storage";
export type { ClientError, GCSMetaStorageOptions, GCStorageOptions } from "./types";
