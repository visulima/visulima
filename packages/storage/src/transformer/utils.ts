import mime from "mime/lite";

import type { FileReturn } from "../storage/utils/file";

/**
 * Check if a content type is valid for a specific media type
 * @param contentType MIME content type string
 * @param expectedType Expected media type ('image', 'video', or 'audio')
 * @returns True if the content type is valid for the expected media type
 */
export const isValidMediaType = (contentType: string | undefined, expectedType: "image" | "video" | "audio"): boolean => {
    if (!contentType) {
        return false;
    }

    return contentType.startsWith(`${expectedType}/`);
};

/**
 * Get format (extension) from content type using mime package
 * @param contentType MIME content type string
 * @returns Format string or undefined if not found
 */
export const getFormatFromContentType = (contentType: string | undefined): string | undefined => {
    if (!contentType) {
        return undefined;
    }

    return mime.getExtension(contentType) || undefined;
};

/**
 * Validate a media file for a specific type
 * @param file File to validate
 * @param expectedType Expected media type
 * @param config Validation configuration
 * @throws Error if validation fails
 */
export const validateMediaFile = (
    file: FileReturn,
    expectedType: "image" | "video" | "audio",
    config?: {
        maxSize?: number;
        supportedFormats?: string[];
    },
): void => {
    // Check file size
    const fileSize = typeof file.size === "string" ? Number.parseInt(file.size, 10) : file.size;

    if (config?.maxSize && fileSize > config.maxSize) {
        throw new Error(`${expectedType} size ${fileSize} exceeds maximum allowed size ${config.maxSize}`);
    }

    // Check if it's the expected media type
    if (!isValidMediaType(file.contentType, expectedType)) {
        throw new Error(`File is not ${expectedType === "image" ? "an" : "a"} ${expectedType}: ${file.contentType}`);
    }

    // Check format support
    const format = getFormatFromContentType(file.contentType);

    if (config?.supportedFormats && format && !config.supportedFormats.includes(format)) {
        throw new Error(`Unsupported ${expectedType} format: ${format}`);
    }
};

/**
 * Validate that a content type is known and supported by the mime package
 * @param contentType MIME content type string
 * @returns True if the content type is known
 */
export const isKnownContentType = (contentType: string | undefined): boolean => {
    if (!contentType) {
        return false;
    }

    return !!mime.getExtension(contentType);
};
