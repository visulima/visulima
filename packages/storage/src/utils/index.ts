import fnv1a from "@sindresorhus/fnv1a";
import mem from "memoize";

export const hash = mem((value: string) => fnv1a(value, { size: 64 }).toString(16));

export type { ErrorResponses } from "./errors";
export { ErrorMap, ERRORS, isUploadError, throwErrorCode, UploadError } from "./errors";
export type { PathMatch } from "./file-path-url-matcher";
export { default as filePathUrlMatcher } from "./file-path-url-matcher";
export {
    appendHeader,
    extractHost,
    extractProto,
    getBaseUrl,
    getHeader,
    getIdFromRequest,
    getMetadata,
    getRealPath,
    getRequestStream,
    normalizeHookResponse,
    normalizeOnErrorResponse,
    readBody,
    setHeaders,
    uuidRegex,
} from "./http";
export { default as Locker } from "./locker";
export { StreamChecksum, streamChecksum } from "./pipes/stream-checksum";
export { default as StreamLength } from "./pipes/stream-length";
export { default as isEqual } from "./primitives/is-equal";
export { default as isRecord } from "./primitives/is-record";
export { default as mapValues } from "./primitives/map-values";
export { default as pick } from "./primitives/pick";
export { default as toMilliseconds } from "./primitives/to-milliseconds";
export { default as toSeconds } from "./primitives/to-seconds";
export { default as RangeHasher } from "./range-hasher";
export type {
    Header,
    Headers,
    HttpError,
    HttpErrorBody,
    IncomingMessageWithBody,
    Logger,
    ResponseBody,
    ResponseBodyType,
    ResponseTuple,
    UploadResponse,
    Validation,
    ValidationError,
    ValidatorConfig,
} from "./types";
export { isValidationError, Validator } from "./validator";
