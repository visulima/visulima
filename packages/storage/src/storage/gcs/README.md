# Google Cloud Storage

Google Cloud Storage (GCS) implementation for the Visulima upload storage system. Provides full GCS compatibility with advanced features like storage classes, lifecycle management, and global CDN integration.

## Features

- ✅ **Full GCS Compatibility**: Complete GCS API support
- ✅ **Storage Classes**: Standard, Nearline, Coldline, Archive
- ✅ **Lifecycle Policies**: Automatic TTL and storage transitions
- ✅ **Signed URLs**: V4 signed URLs for secure access
- ✅ **Global CDN**: Worldwide content delivery
- ✅ **Versioning**: Object versioning support
- ✅ **Custom Metadata**: Rich metadata support
- ✅ **Resumable Uploads**: Large file upload support

## Installation

```bash
npm install @google-cloud/storage
```

## Basic Usage

```typescript
import { GCStorage } from "@visulima/upload/storage";

const storage = new GCStorage({
    bucket: "my-upload-bucket",
    credentials: {
        client_email: "...",
        client_id: "...",
        private_key: "...",
        private_key_id: "...",
        project_id: "my-project-id",
        type: "service_account",
    },
    projectId: "my-project-id",
});

// Use storage...
const file = await storage.create(request, {
    contentType: "image/jpeg",
    metadata: { filename: "photo.jpg" },
});
```

## Configuration

### Service Account Key

```typescript
const storage = new GCStorage({
    bucket: "my-bucket",
    keyFilename: "/path/to/service-account-key.json",
    projectId: "my-project-123",
});
```

### Environment Variables

```typescript
// Set GOOGLE_APPLICATION_CREDENTIALS
process.env.GOOGLE_APPLICATION_CREDENTIALS = "/path/to/key.json";

const storage = new GCStorage({
    bucket: "my-bucket",
    projectId: "my-project-123",
});
```

### Advanced Configuration

```typescript
const storage = new GCStorage({
    bucket: "my-bucket",
    credentials: {
        /* service account */
    },
    // Generic storage config
    genericConfig: {
        checksumTypes: ["md5", "crc32c"],
        maxFileSize: 5 * 1024 * 1024 * 1024 * 1024, // 5TB
        optimizations: {
            cacheStorageClass: "NEARLINE",
            enableCompression: false,
            prefixTemplate: "uploads/{fileId}/",
            usePrefixes: true,
        },
    },

    projectId: "my-project-123",

    // GCS-specific options
    retryOptions: {
        maxRetryDelay: 64_000,
        retryableErrorFn: (error) => error.code !== 404,
        retryDelayMultiplier: 2,
        totalTimeout: 600_000, // 10 minutes
    },
});
```

## Storage Features

Google Cloud Storage provides the following features:

- **File Operations**: Upload, download, delete, copy, move
- **Metadata Support**: Custom metadata and system metadata
- **Checksum Validation**: MD5, CRC32C
- **Storage Classes**: Standard, Nearline, Coldline, Archive
- **Lifecycle Policies**: Automatic TTL and class transitions
- **Signed URLs**: V4 signed URLs for secure access
- **Resumable Uploads**: Large file uploads with continuation
- **Server-side Copy**: Efficient file copying within GCS
- **Conditional Operations**: ETags and conditional headers
- **Maximum File Size**: 5TB per file
- **Maximum Part Size**: 2GB per part

## Storage Classes

GCS offers different storage classes for cost optimization:

```typescript
// Set storage class during copy
await storage.copy(file.id, `archive/${file.id}`, { storageClass: "COLDLINE" });
```

## TTL (Time-to-Live)

Set expiration times for files:

```typescript
// Create file with TTL
await storage.create(request, {
    contentType: "image/jpeg",
    metadata: { filename: "temp.jpg" },
    ttl: "30d", // Expires in 30 days
});

// Update file TTL
await storage.update({ id: file.id }, { ttl: 24 * 60 * 60 * 1000 }); // 24 hours
```

## Lifecycle Management

Configure automatic lifecycle policies:

```typescript
// Set TTL (moves to cheaper storage automatically)
await storage.setTTL({ id: file.id }, 30 * 24 * 60 * 60 * 1000); // 30 days

// GCS lifecycle rules can be configured in the console or via API
// to automatically transition files between storage classes
```

## Signed URLs

Generate V4 signed URLs for secure access:

```typescript
// Generate signed URL for download (1 hour)
const downloadUrl = await storage.getUrl({ id: file.id }, 60 * 60 * 1000);

// Generate signed URL for upload
const uploadUrl = await storage.getUploadUrl({ id: file.id }, 60 * 60 * 1000);
```

## Resumable Uploads

GCS supports resumable uploads for large files:

```typescript
// Large file upload (automatically resumable)
const file = await storage.create(request, {
    contentType: "video/mp4",
    size: 10 * 1024 * 1024 * 1024, // 10GB
});

// The storage will automatically:
// - Use resumable upload protocol
// - Handle network interruptions
// - Resume from last successful chunk
// - Provide upload progress
```

## Custom Metadata

Rich metadata support:

```typescript
const file = await storage.create(request, {
    contentType: "image/jpeg",
    metadata: {
        author: "John Doe",
        customField: "custom-value",
        tags: ["vacation", "beach"],
    },
});

// Update metadata
await storage.update(
    { id: file.id },
    {
        metadata: { ...file.metadata, processed: true },
    },
);
```

## Versioning

Enable object versioning for backup and recovery:

```typescript
// Enable versioning in GCS console or via API
// Then access specific versions
const versions = await storage.listVersions(fileId);
const latestVersion = await storage.get({ generation: versionId, id: fileId });
```

## CDN Integration

GCS integrates seamlessly with Google Cloud CDN:

```typescript
const storage = new GCStorage({
    bucket: "my-bucket",
    genericConfig: {
        optimizations: {
            cacheStorageClass: "NEARLINE",
            enableCDNHeaders: true, // Add CDN-friendly headers
        },
    },
});

// Copy with storage class
await storage.copy("file1", "archive/file1", {
    storageClass: "COLDLINE",
});

// Files served through CDN automatically
// Configure CDN in Google Cloud Console
```

## Error Handling

GCS-specific errors are properly mapped:

```typescript
try {
    await storage.get({ id: "nonexistent" });
} catch (error) {
    console.log(error.code); // 'Not Found'
    console.log(error.statusCode); // 404
    console.log(error.message); // 'The specified key does not exist.'
}
```

## Global Distribution

GCS provides worldwide data distribution:

### Multi-region Buckets

```typescript
// Create multi-region bucket for global distribution
const storage = new GCStorage({
    bucket: "my-global-bucket",
    location: "us", // Multi-region
});

// Automatic replication across regions
// Lower latency for global users
```

### Regional Buckets

```typescript
// Regional bucket for single region performance
const storage = new GCStorage({
    bucket: "my-regional-bucket",
    location: "us-central1",
});
```

## Performance Optimization

### Upload Optimization

```typescript
const storage = new GCStorage({
    bucket: "my-bucket",
    // Increase chunk size for faster uploads
    chunkSize: 8 * 1024 * 1024, // 8MB
    // Optimize for large files
    retryOptions: {
        maxRetries: 3,
        retryDelayMultiplier: 1.5,
    },
});
```

### Download Optimization

```typescript
// Use signed URLs for direct downloads
const signedUrl = await storage.getUrl({ id: file.id }, 3_600_000);
// Users download directly from GCS, not through your server
```

## Environment Variables

```bash
# Service Account
GOOGLE_APPLICATION_CREDENTIALS=/path/to/service-account-key.json

# Or explicit credentials
GCS_PROJECT_ID=my-project-id
GCS_BUCKET=my-bucket
GCS_KEY_FILENAME=/path/to/key.json

# Optional
GCS_STORAGE_API_VERSION=v1
GCS_RETRY_LIMIT=3
```

## Monitoring and Metrics

GCS provides detailed metrics and monitoring:

```typescript
// Enable detailed logging
const storage = new GCStorage({
    bucket: "my-bucket",
    logger: console,
});

// GCS metrics available in Google Cloud Console:
// - Request count, latency, error rates
// - Storage usage by class
// - Network egress
// - API costs
```

## Testing

Run GCS-specific tests:

```bash
# Run GCS tests
npm test -- --testPathPattern="gcs.*test"

# With real GCS
GOOGLE_APPLICATION_CREDENTIALS=/path/to/key.json npm test
```

## Troubleshooting

### Authentication Issues

```typescript
// Verify credentials
const storage = new GCStorage({
    bucket: "test-bucket",
    projectId: "test-project",
});

// Test authentication
try {
    await storage.list(1);
    console.log("Authentication successful");
} catch (error) {
    console.error("Authentication failed:", error.message);
    // Check service account permissions
    // Ensure Storage Object Admin role
}
```

### Permission Issues

Ensure your service account has appropriate permissions:

- `Storage Object Admin` - Full control
- `Storage Object Creator` - Upload only
- `Storage Object Viewer` - Read only

### CORS Configuration

For web applications, configure CORS in GCS:

```bash
# Use gsutil or Cloud Console
gsutil cors set cors-config.json gs://my-bucket

# cors-config.json
[
  {
    "origin": ["https://myapp.com"],
    "method": ["GET", "POST", "PUT"],
    "responseHeader": ["Content-Type"],
    "maxAgeSeconds": 3600
  }
]
```

### Quota Limits

Be aware of GCS quotas:

- 1000 requests/second per bucket
- 5TB/day upload limit (can be increased)
- Rate limits vary by operation type

For high-traffic applications, consider:

- Multiple buckets
- Regional buckets for lower latency
- CDN integration to reduce origin requests
