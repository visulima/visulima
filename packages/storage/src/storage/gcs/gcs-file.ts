import { File } from "../utils/file";

class GCSFile extends File {
    public GCSUploadURI?: string;

    public uri?: string;
}

export default GCSFile;
