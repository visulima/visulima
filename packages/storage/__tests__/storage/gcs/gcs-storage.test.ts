import type { Response as NodeFetchResponse } from "node-fetch";
import fetch from "node-fetch";
import { createRequest } from "node-mocks-http";
import { afterEach, beforeEach, describe, expect, expectTypeOf, it, vi } from "vitest";

import FetchError from "../../../src/storage/gcs/fetch-error";
import type GCSFile from "../../../src/storage/gcs/gcs-file";
import GCStorage from "../../../src/storage/gcs/gcs-storage";
import type { ClientError, GCStorageOptions } from "../../../src/storage/gcs/types";
import { buildContentRange, getRangeEnd } from "../../../src/storage/gcs/utils";
import type { FilePart } from "../../../src/storage/utils/file";
import { metafile, storageOptions, testfile } from "../../__helpers__/config";
import { deepClone } from "../../__helpers__/utils";

const { mockFetch } = vi.hoisted(() => {
    return {
        mockFetch: vi.fn(),
    };
});

vi.mock(import("node-fetch"), () => {
    return {
        default: mockFetch,
    };
});

const mockAuthRequest = vi.fn();

vi.mock(import("google-auth-library"), () => {
    return {
        GoogleAuth: vi.fn(() => {
            return { request: mockAuthRequest };
        }),
    };
});

describe(GCStorage, async () => {
    const { Response } = await vi.importActual<{ Response: typeof NodeFetchResponse }>("node-fetch");

    vi.useFakeTimers().setSystemTime(new Date("2022-02-02"));

    let storage: GCStorage;

    // eslint-disable-next-line no-secrets/no-secrets,radar/no-duplicate-string
    const uri = "http://api.com?upload_id=123456789";
    const request = createRequest({ headers: { origin: "http://api.com" } });
    const metafileResponse = (): { data: GCSFile; status: number } =>
        deepClone({
            data: { ...metafile, createdAt: new Date().toISOString(), uri },
            status: 200,
        });

    beforeEach(async () => {
        vi.clearAllMocks();

        mockAuthRequest.mockResolvedValueOnce({ bucket: "ok" });

        storage = new GCStorage({ ...(storageOptions as GCStorageOptions), bucket: "test-bucket", projectId: "test" });
    });

    afterEach(() => {
        vi.clearAllMocks();
    });

    describe(".create()", () => {
        it("should make API request and set file status and URI", async () => {
            expect.assertions(2);

            mockAuthRequest.mockRejectedValueOnce({ code: 404, detail: "meta not found" }); // getMeta
            mockAuthRequest.mockResolvedValueOnce({
                headers: { get: (name: string) => (name === "location" ? uri : name === "X-Goog-Upload-Status" ? "active" : undefined) },
                status: 200,
            }); // create
            mockAuthRequest.mockResolvedValueOnce("_saveOk"); // saveMeta

            const gcsFile = await storage.create(request, metafile);

            expect(gcsFile).toMatchSnapshot();
            expect(mockAuthRequest).toMatchSnapshot();
        });

        it("should handle existing files correctly", async () => {
            expect.assertions(1);

            mockAuthRequest.mockResolvedValue(metafileResponse());
            mockFetch.mockResolvedValueOnce(new Response("", { headers: { Range: "0-5" }, status: 308 }));

            const gcsFile = await storage.create(request, metafile);

            expect(gcsFile).toMatchSnapshot();
        });

        it("should reject when API returns an error", async () => {
            expect.assertions(1);

            const errorObject = { code: 403, response: {} };

            mockAuthRequest.mockRejectedValue(errorObject);

            await expect(storage.create(request, metafile)).rejects.toEqual(errorObject);
        });

        it("should handle TTL option and set expiration timestamp", async () => {
            mockAuthRequest.mockRejectedValueOnce({ code: 404, detail: "meta not found" }); // getMeta
            mockAuthRequest.mockResolvedValueOnce({
                headers: { get: (name: string) => (name === "location" ? uri : name === "X-Goog-Upload-Status" ? "active" : undefined) },
                status: 200,
            }); // create
            mockAuthRequest.mockResolvedValueOnce("_saveOk"); // saveMeta

            const gcsFile = await storage.create(request, { ...metafile, ttl: "7d" });

            expect(gcsFile.expiredAt).toBeDefined();

            expectTypeOf(gcsFile.expiredAt).toBeNumber();

            // TTL should be converted to expiredAt timestamp
            const expectedExpiry = Date.now() + 7 * 24 * 60 * 60 * 1000; // 7 days in ms

            expect(gcsFile.expiredAt).toBeGreaterThan(expectedExpiry - 1000); // Allow 1s tolerance
            expect(gcsFile.expiredAt).toBeLessThan(expectedExpiry + 1000);
        });
    });

    describe(".update()", () => {
        it("should update changed metadata keys correctly", async () => {
            expect.assertions(3);

            mockAuthRequest.mockResolvedValue(metafileResponse());

            // eslint-disable-next-line radar/no-duplicate-string
            const gcsFile = await storage.update(metafile, { metadata: { name: "newname.mp4" } });

            expect(gcsFile.metadata.name).toBe("newname.mp4");
            expect(gcsFile.originalName).toBe("newname.mp4");
            expect(gcsFile.metadata.mimeType).toBe("video/mp4");
        });

        it("should reject update operation when file is not found", async () => {
            expect.assertions(1);

            mockAuthRequest.mockResolvedValue({});

            await expect(storage.update(metafile, { metadata: { name: "newname.mp4" } })).rejects.toHaveProperty("UploadErrorCode", "FileNotFound");
        });

        it("should handle TTL option and set expiration timestamp during update", async () => {
            mockAuthRequest.mockResolvedValue(metafileResponse());

            const gcsFile = await storage.update(metafile, { ttl: "2h" });

            expect(gcsFile.expiredAt).toBeDefined();

            expectTypeOf(gcsFile.expiredAt).toBeNumber();

            // TTL should be converted to expiredAt timestamp
            const expectedExpiry = Date.now() + 2 * 60 * 60 * 1000; // 2 hours in ms

            expect(gcsFile.expiredAt).toBeGreaterThan(expectedExpiry - 1000); // Allow 1s tolerance
            expect(gcsFile.expiredAt).toBeLessThan(expectedExpiry + 1000);
        });
    });

    describe(".write()", () => {
        it("should make API request and set file status and bytesWritten", async () => {
            expect.assertions(3);

            vi.spyOn(storage, "getMeta").mockResolvedValue(metafile);

            mockAuthRequest.mockResolvedValueOnce({ data: { mediaLink: uri }, status: 200 });

            mockFetch.mockResolvedValueOnce(new Response("{\"mediaLink\":\"http://api.com/123456789\"}", { status: 200 }));

            const body = testfile.asReadable;
            const gcsFile = await storage.write({
                body,
                contentLength: metafile.size,
                id: metafile.id,
                start: 0,
            });

            expect(mockFetch).toHaveBeenCalledTimes(1);
            expect(mockFetch).toHaveBeenCalledWithExactlyOnceWith(uri, {
                body,
                headers: expect.objectContaining({ "Content-Range": "bytes 0-63/64" }),
                method: "PUT",
                signal: expect.any(AbortSignal),
            });
            expect(gcsFile).toMatchSnapshot();
        });

        it("should send normalized error for API failures", async () => {
            expect.assertions(2);

            mockAuthRequest.mockResolvedValueOnce(metafileResponse());

            // Mock the makeRequest to return error response
            mockAuthRequest.mockResolvedValueOnce({ data: "Bad Request", status: 400 });

            try {
                await storage.write({ contentLength: 0, id: metafile.id });
            } catch (error) {
                // eslint-disable-next-line no-secrets/no-secrets
                expect(error).toEqual(new FetchError("Bad Request", "GCS400", { uri: "http://api.com?upload_id=123456789" }));
                expect(error).toMatchSnapshot();
            }
        });

        it("should make API request and set status and bytesWritten when resuming", async () => {
            expect.assertions(3);

            vi.spyOn(storage, "getMeta").mockResolvedValue({ ...metafile, bytesWritten: 0 });

            mockAuthRequest.mockResolvedValueOnce({ data: { mediaLink: uri }, status: 200 });

            mockFetch.mockResolvedValueOnce(new Response("", { headers: { Range: "0-5" }, status: 200 }));

            const gcsFile = await storage.write({ contentLength: 0, id: metafile.id });

            expect(mockFetch).toMatchSnapshot();
            expect(gcsFile.status).toBe("part");
            expect(gcsFile.bytesWritten).toBe(6);
        });
    });

    describe(".delete()", () => {
        it("should mark file as deleted and return file data", async () => {
            expect.assertions(2);

            mockAuthRequest.mockResolvedValue({ data: { ...metafile, uri } });

            const deleted = await storage.delete(metafile);

            expect(deleted.id).toBe(metafile.id);
            expect(deleted.status).toBe("deleted");
        });

        it("should handle deletion of non-existent files gracefully", async () => {
            expect.assertions(1);

            mockAuthRequest.mockResolvedValue({});

            const deleted = await storage.delete(metafile);

            expect(deleted.id).toBe(metafile.id);
        });
    });

    describe(".copy()", () => {
        it("should copy file to relative path with correct API calls", async () => {
            expect.assertions(5);

            mockAuthRequest.mockClear();
            mockAuthRequest.mockResolvedValue({ data: { done: true } });

            await storage.copy(metafile, "files/новое имя.txt");

            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledWith({
                url: "https://storage.googleapis.com/storage/v1/b/test-bucket",
            });

            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledWith({
                body: "",
                headers: { "Content-Type": "application/json" },
                method: "POST",
                // eslint-disable-next-line no-secrets/no-secrets
                url: "https://storage.googleapis.com/storage/v1/b/test-bucket/o/testfile.mp4/rewriteTo/b/test-bucket/o/files/%D0%BD%D0%BE%D0%B2%D0%BE%D0%B5%20%D0%B8%D0%BC%D1%8F.txt",
            });
        });

        it("should copy file to absolute path with correct API calls", async () => {
            expect.assertions(3);

            mockAuthRequest.mockClear();
            mockAuthRequest.mockResolvedValue({ data: { done: true } });

            await storage.copy(metafile, "/new/name.txt");

            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledWith({
                body: "",
                headers: { "Content-Type": "application/json" },
                method: "POST",
                url: "https://storage.googleapis.com/storage/v1/b/test-bucket/o/testfile.mp4/rewriteTo/b/new/o/name.txt",
            });
        });

        it("should copy file with storage class option", async () => {
            expect.assertions(3);

            mockAuthRequest.mockClear();
            mockAuthRequest.mockResolvedValue({ data: { done: true } });

            await storage.copy(metafile, "files/backup.txt", { storageClass: "COLDLINE" });

            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledTimes(1);
            expect(mockAuthRequest).toHaveBeenCalledWith({
                body: "",
                headers: { "Content-Type": "application/json" },
                method: "POST",
                url: "https://storage.googleapis.com/storage/v1/b/test-bucket/o/testfile.mp4/rewriteTo/b/test-bucket/o/files/backup.txt",
            });
        });
    });

    describe("normalizeError", () => {
        it("should normalize client error with correct code and status", () => {
            expect.assertions(1);

            const error: ClientError = {
                code: "400",
                // eslint-disable-next-line no-secrets/no-secrets
                config: { uri: "http://api.com?upload_id=123456789" },
                message: "Bad Request",
                name: "ClientError",
            };

            expect(storage.normalizeError(error)).toEqual(expect.objectContaining({ code: "GCS400", statusCode: 400 }));
        });

        it("should normalize non-client errors with generic error code", () => {
            expect.assertions(1);

            expect(storage.normalizeError(new Error("unknown") as ClientError)).toEqual(expect.objectContaining({ code: "GenericUploadError" }));
        });
    });
});

describe("range utils", () => {
    it.each([
        ["", 0],
        ["0-0", 0],
        ["0-1", 2],
        ["0-10000", 10_001],
    ])("should calculate correct range end for input: %s -> %i", (string_, expected) => {
        expect.assertions(1);

        expect(getRangeEnd(string_)).toBe(expected);
    });

    const body = true;

    it.each([
        [{}, "bytes */*"],
        [{ body }, "bytes */*"],
        [{ start: 0 }, "bytes */*"],
        [{ body, start: 0 }, "bytes 0-*/*"],
        [{ body, size: 80, start: 10 }, "bytes 10-*/80"],
        [
            {
                body,
                contentLength: 80,
                size: 80,
                start: 0,
            },
            "bytes 0-79/80",
        ],
        [{ contentLength: 80, size: 80, start: 0 }, "bytes */80"],
    ])("should build correct content range header for input: %o -> %s", (string_, expected) => {
        expect.assertions(1);

        expect(buildContentRange(string_ as FilePart & GCSFile)).toBe(expected);
    });
});
