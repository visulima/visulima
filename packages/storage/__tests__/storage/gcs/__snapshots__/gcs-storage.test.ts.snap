// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gcs > .create() > should handle existing files correctly 1`] = `
{
  "bytesWritten": 64,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "uri": "http://api.com?upload_id=123456789",
}
`;

exports[`gcs > .create() > should make API request and set file status and URI 1`] = `
GCSFile {
  "ETag": undefined,
  "GCSUploadURI": undefined,
  "bytesWritten": 0,
  "content": undefined,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": 1643763600000,
  "hash": undefined,
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "modifiedAt": undefined,
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
  "uri": "http://api.com?upload_id=123456789",
}
`;

exports[`gcs > .create() > should make API request and set file status and URI 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "headers": {
          "User-Agent": "@visulima/upload/0.0.0",
          "x-goog-api-client": "gl-node/24.9.0 gccl/0.0.0 gccl-invocation-id/0d798c61-8e37-4407-b100-174afb52750d",
        },
        "params": {},
        "retry": true,
        "retryConfig": {
          "retry": 3,
          "shouldRetry": [Function],
          "statusCodesToRetry": [
            [
              408,
              429,
              500,
              502,
              503,
              504,
            ],
            [
              100,
              199,
            ],
            [
              429,
            ],
            [
              500,
              599,
            ],
          ],
        },
        "timeout": 60000,
        "url": "https://storage.googleapis.com/storage/v1/b/test-bucket",
      },
    ],
    [
      {
        "headers": {
          "User-Agent": "@visulima/upload/0.0.0",
          "x-goog-api-client": "gl-node/24.9.0 gccl/0.0.0 gccl-invocation-id/2a886d76-1d86-41d4-8539-5f1cb3c33646",
        },
        "params": {},
        "retry": true,
        "retryConfig": {
          "retry": 3,
          "shouldRetry": [Function],
          "statusCodesToRetry": [
            [
              408,
              429,
              500,
              502,
              503,
              504,
            ],
            [
              100,
              199,
            ],
            [
              429,
            ],
            [
              500,
              599,
            ],
          ],
        },
        "timeout": 60000,
        "url": "https://storage.googleapis.com/storage/v1/b/test-bucket/o/e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78.META",
      },
    ],
    [
      {
        "body": "{"metadata":{"custom":"","lastModified":1635398061454,"mimeType":"video/mp4","name":"testfile.mp4","sha1":"ZAPAntzKARqtb+j3B529GAOf3kI=","size":64}}",
        "headers": {
          "User-Agent": "@visulima/upload/0.0.0",
          "x-goog-api-client": "gl-node/24.9.0 gccl/0.0.0 gccl-invocation-id/d4e7a56b-18bb-4c2a-981d-ddc71e61334d",
        },
        "method": "POST",
        "params": {},
        "retry": true,
        "retryConfig": {
          "retry": 3,
          "shouldRetry": [Function],
          "statusCodesToRetry": [
            [
              408,
              429,
              500,
              502,
              503,
              504,
            ],
            [
              100,
              199,
            ],
            [
              429,
            ],
            [
              500,
              599,
            ],
          ],
        },
        "timeout": 60000,
        "url": "https://storage.googleapis.com/upload/storage/v1/b/test-bucket/o",
      },
    ],
    [
      {
        "body": "{"bytesWritten":0,"contentType":"video/mp4","originalName":"testfile.mp4","id":"e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78","metadata":"custom ,lastModified MTYzNTM5ODA2MTQ1NA==,mimeType dmlkZW8vbXA0,name dGVzdGZpbGUubXA0,sha1 WkFQQW50ektBUnF0YitqM0I1MjlHQU9mM2tJPQ==,size NjQ=","name":"anonymous/testfile.mp4","size":64,"expiredAt":1643763600000,"createdAt":"2022-02-02T00:00:00.000Z","uri":"http://api.com?upload_id=123456789"}",
        "headers": {
          "User-Agent": "@visulima/upload/0.0.0",
          "x-goog-api-client": "gl-node/24.9.0 gccl/0.0.0 gccl-invocation-id/3a0f4d6c-f618-425b-bddb-fe7b20c536dc",
        },
        "method": "POST",
        "params": {},
        "retry": true,
        "retryConfig": {
          "retry": 3,
          "shouldRetry": [Function],
          "statusCodesToRetry": [
            [
              408,
              429,
              500,
              502,
              503,
              504,
            ],
            [
              100,
              199,
            ],
            [
              429,
            ],
            [
              500,
              599,
            ],
          ],
        },
        "timeout": 60000,
        "url": "https://storage.googleapis.com/upload/storage/v1/b/test-bucket/o",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`gcs > .write() > should make API request and set status and bytesWritten when resuming 1`] = `[MockFunction spy]`;
